### PR TITLE
etcd-tester: close leaky gRPC connections

### PR DESCRIPTION
**Pull request by gyuho**
For https://github.com/coreos/etcd/issues/4465

### CODE REVIEW

**Comment by gyuho** [comment]
/cc @xiang90 after this, the leaky goroutines decrease from 80 to 14 or less. We still have leaky goroutines from web server (but we can ignore this) but most leaky ones are from gRPC sleeps in resetTransport:

```
=== RUN   TestRun
grpc sleeping... 999.721032ms
grpc sleeping... 999.510505ms
2016/02/09 14:53:53 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:53:53 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:53:54 etcd-tester: [round#0 case#0] start failure kill all members
2016/02/09 14:53:54 etcd-tester: [round#0 case#0] start injecting failure...
2016/02/09 14:53:54 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:53:54 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:53:54 transport: http2Client.notifyError got notified that the client transport was broken read tcp 127.0.0.1:41906->127.0.0.1:2378: read: connection reset by peer.
2016/02/09 14:53:54 etcd-tester: [round#0 case#0] injected failure
2016/02/09 14:53:54 etcd-tester: [round#0 case#0] start recovering failure...
grpc sleeping... 999.788015ms
grpc sleeping... 999.910458ms
grpc sleeping... 999.835007ms
grpc sleeping... 999.907084ms
2016/02/09 14:53:55 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:53:55 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:53:55 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:53:55 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:53:56 etcd-tester: [round#0 case#0] recovered failure
2016/02/09 14:53:56 etcd-tester: [round#0 case#0] canceling the stressers...
2016/02/09 14:53:56 etcd-tester: [round#0 case#0] canceled stressers
2016/02/09 14:53:56 etcd-tester: [round#0 case#0] checking current revisions...
2016/02/09 14:53:57 etcd-tester: [round#0 case#0] all members are consistent with current revisions
2016/02/09 14:53:57 etcd-tester: [round#0 case#0] checking current storage hashes...
2016/02/09 14:53:57 etcd-tester: [round#0 case#0] all members are consistent with storage hashes
2016/02/09 14:53:57 etcd-tester: [round#0 case#0] restarting the stressers...
2016/02/09 14:53:57 etcd-tester: [round#0 case#0] succeed!
2016/02/09 14:53:57 etcd-tester: [round#0 case#1] start failure kill majority of the cluster
2016/02/09 14:53:57 etcd-tester: [round#0 case#1] start injecting failure...
2016/02/09 14:53:57 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:53:57 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:53:57 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:53:57 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:53:57 etcd-tester: [round#0 case#1] injected failure
2016/02/09 14:53:57 etcd-tester: [round#0 case#1] start recovering failure...
2016/02/09 14:53:57 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:53:57 transport: http2Client.notifyError got notified that the client transport was broken EOF.
grpc sleeping... 999.915841ms
grpc sleeping... 999.89934ms
grpc sleeping... 999.807034ms
grpc sleeping... 999.690391ms
grpc sleeping... 999.865601ms
grpc sleeping... 999.656402ms
grpc sleeping... 999.899398ms
2016/02/09 14:53:58 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:53:58 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:53:58 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:53:58 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:53:58 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:53:58 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:53:58 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:53:59 etcd-tester: [round#0 case#1] recovered failure
2016/02/09 14:53:59 etcd-tester: [round#0 case#1] canceling the stressers...
2016/02/09 14:53:59 etcd-tester: [round#0 case#1] canceled stressers
2016/02/09 14:53:59 etcd-tester: [round#0 case#1] checking current revisions...
2016/02/09 14:54:00 etcd-tester: [round#0 case#1] all members are consistent with current revisions
2016/02/09 14:54:00 etcd-tester: [round#0 case#1] checking current storage hashes...
2016/02/09 14:54:00 etcd-tester: [round#0 case#1] all members are consistent with storage hashes
2016/02/09 14:54:00 etcd-tester: [round#0 case#1] restarting the stressers...
2016/02/09 14:54:00 etcd-tester: [round#0 case#1] succeed!
2016/02/09 14:54:00 etcd-tester: [round#0 case#2] start failure kill one random member
2016/02/09 14:54:00 etcd-tester: [round#0 case#2] start injecting failure...
2016/02/09 14:54:00 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:00 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:00 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:00 transport: http2Client.notifyError got notified that the client transport was broken read tcp 127.0.0.1:41939->127.0.0.1:2378: read: connection reset by peer.
2016/02/09 14:54:00 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:00 transport: http2Client.notifyError got notified that the client transport was broken EOF.
grpc sleeping... 999.613858ms
2016/02/09 14:54:00 etcd-tester: [round#0 case#2] injected failure
2016/02/09 14:54:00 etcd-tester: [round#0 case#2] start recovering failure...
grpc sleeping... 999.750779ms
2016/02/09 14:54:00 transport: http2Client.notifyError got notified that the client transport was broken EOF.
grpc sleeping... 998.80898ms
grpc sleeping... 999.056828ms
grpc sleeping... 998.78252ms
2016/02/09 14:54:00 transport: http2Client.notifyError got notified that the client transport was broken EOF.
grpc sleeping... 998.998009ms
2016/02/09 14:54:00 transport: http2Client.notifyError got notified that the client transport was broken EOF.
grpc sleeping... 999.895738ms
grpc sleeping... 999.809448ms
grpc sleeping... 999.893906ms
grpc sleeping... 999.911512ms
2016/02/09 14:54:01 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:54:01 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:54:01 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:54:01 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:54:01 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:54:01 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:54:01 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:54:01 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:54:01 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:54:01 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:54:02 etcd-tester: [round#0 case#2] recovered failure
2016/02/09 14:54:02 etcd-tester: [round#0 case#2] canceling the stressers...
2016/02/09 14:54:02 etcd-tester: [round#0 case#2] canceled stressers
2016/02/09 14:54:02 etcd-tester: [round#0 case#2] checking current revisions...
2016/02/09 14:54:03 etcd-tester: [round#0 case#2] all members are consistent with current revisions
2016/02/09 14:54:03 etcd-tester: [round#0 case#2] checking current storage hashes...
2016/02/09 14:54:03 etcd-tester: [round#0 case#2] all members are consistent with storage hashes
2016/02/09 14:54:03 etcd-tester: [round#0 case#2] restarting the stressers...
2016/02/09 14:54:03 etcd-tester: [round#0 case#2] succeed!
2016/02/09 14:54:03 etcd-tester: [round#0 case#3] start failure kill one member for long time and expect it to recover from incoming snapshot
2016/02/09 14:54:03 etcd-tester: [round#0 case#3] start injecting failure...
2016/02/09 14:54:03 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:03 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:03 transport: http2Client.notifyError got notified that the client transport was broken EOF.
grpc sleeping... 999.850073ms
2016/02/09 14:54:03 etcd-tester: [round#0 case#3] injected failure
2016/02/09 14:54:03 etcd-tester: [round#0 case#3] start recovering failure...
grpc sleeping... 999.760343ms
2016/02/09 14:54:03 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:03 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:03 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:03 transport: http2Client.notifyError got notified that the client transport was broken EOF.
grpc sleeping... 999.867365ms
2016/02/09 14:54:03 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:03 transport: http2Client.notifyError got notified that the client transport was broken EOF.
grpc sleeping... 999.908787ms
grpc sleeping... 999.889471ms
grpc sleeping... 999.902616ms
grpc sleeping... 999.910699ms
grpc sleeping... 999.912744ms
2016/02/09 14:54:03 transport: http2Client.notifyError got notified that the client transport was broken EOF.
grpc sleeping... 999.903061ms
2016/02/09 14:54:03 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:03 transport: http2Client.notifyError got notified that the client transport was broken read tcp 127.0.0.1:41965->127.0.0.1:2378: read: connection reset by peer.
grpc sleeping... 999.902898ms
grpc sleeping... 999.917222ms
grpc sleeping... 999.863911ms
grpc sleeping... 999.890789ms
2016/02/09 14:54:04 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:54:04 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:54:04 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:54:04 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:54:04 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:54:04 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:54:04 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:54:04 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:54:04 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:54:04 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:54:04 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:54:04 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:54:04 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "localhost:2378"
2016/02/09 14:54:05 etcd-tester: [round#0 case#3] recovered failure
2016/02/09 14:54:05 etcd-tester: [round#0 case#3] canceling the stressers...
2016/02/09 14:54:05 etcd-tester: [round#0 case#3] canceled stressers
2016/02/09 14:54:05 etcd-tester: [round#0 case#3] checking current revisions...
2016/02/09 14:54:06 etcd-tester: [round#0 case#3] all members are consistent with current revisions
2016/02/09 14:54:06 etcd-tester: [round#0 case#3] checking current storage hashes...
2016/02/09 14:54:06 etcd-tester: [round#0 case#3] all members are consistent with storage hashes
2016/02/09 14:54:06 etcd-tester: [round#0 case#3] restarting the stressers...
2016/02/09 14:54:06 etcd-tester: [round#0 case#3] succeed!
2016/02/09 14:54:07 etcd-tester: [round#0 case#4] start failure isolate one member
2016/02/09 14:54:07 etcd-tester: [round#0 case#4] start injecting failure...
2016/02/09 14:54:07 etcd-tester: [round#0 case#4] injected failure
2016/02/09 14:54:07 etcd-tester: [round#0 case#4] start recovering failure...
2016/02/09 14:54:07 etcd-tester: [round#0 case#4] recovered failure
2016/02/09 14:54:07 etcd-tester: [round#0 case#4] canceling the stressers...
2016/02/09 14:54:07 etcd-tester: [round#0 case#4] canceled stressers
2016/02/09 14:54:07 etcd-tester: [round#0 case#4] checking current revisions...
2016/02/09 14:54:08 etcd-tester: [round#0 case#4] all members are consistent with current revisions
2016/02/09 14:54:08 etcd-tester: [round#0 case#4] checking current storage hashes...
2016/02/09 14:54:08 etcd-tester: [round#0 case#4] all members are consistent with storage hashes
2016/02/09 14:54:08 etcd-tester: [round#0 case#4] restarting the stressers...
2016/02/09 14:54:08 etcd-tester: [round#0 case#4] succeed!
2016/02/09 14:54:08 etcd-tester: [round#0 case#5] start failure isolate all members
2016/02/09 14:54:08 etcd-tester: [round#0 case#5] start injecting failure...
2016/02/09 14:54:08 etcd-tester: [round#0 case#5] injected failure
2016/02/09 14:54:08 etcd-tester: [round#0 case#5] start recovering failure...
2016/02/09 14:54:08 etcd-tester: [round#0 case#5] recovered failure
2016/02/09 14:54:08 etcd-tester: [round#0 case#5] canceling the stressers...
2016/02/09 14:54:08 etcd-tester: [round#0 case#5] canceled stressers
2016/02/09 14:54:08 etcd-tester: [round#0 case#5] checking current revisions...
2016/02/09 14:54:09 etcd-tester: [round#0 case#5] all members are consistent with current revisions
2016/02/09 14:54:09 etcd-tester: [round#0 case#5] checking current storage hashes...
2016/02/09 14:54:09 etcd-tester: [round#0 case#5] all members are consistent with storage hashes
2016/02/09 14:54:09 etcd-tester: [round#0 case#5] restarting the stressers...
2016/02/09 14:54:09 etcd-tester: [round#0 case#5] succeed!
2016/02/09 14:54:09 etcd-tester: [round#0] compacting storage at 1726 (current revision 11726)
2016/02/09 14:54:09 etcd-tester: [round#0] compacted storage
2016/02/09 14:54:14 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:14 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:14 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:14 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:14 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:14 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:14 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:14 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:14 transport: http2Client.notifyError got notified that the client transport was broken EOF.
grpc sleeping... 999.900812ms
2016/02/09 14:54:14 transport: http2Client.notifyError got notified that the client transport was broken EOF.
grpc sleeping... 999.802852ms
grpc sleeping... 999.569475ms
grpc sleeping... 999.699215ms
grpc sleeping... 999.553792ms
grpc sleeping... 999.610364ms
2016/02/09 14:54:14 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:14 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:14 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:14 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:14 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:14 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:14 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016/02/09 14:54:14 transport: http2Client.notifyError got notified that the client transport was broken EOF.
grpc sleeping... 999.9294ms
grpc sleeping... 999.822595ms
grpc sleeping... 999.766509ms
grpc sleeping... 999.6674ms
grpc sleeping... 999.338651ms
grpc sleeping... 999.414308ms
grpc sleeping... 999.312283ms
grpc sleeping... 999.483187ms
grpc sleeping... 999.684075ms
grpc sleeping... 999.572315ms
grpc sleeping... 999.553119ms
grpc sleeping... 999.104253ms
--- PASS: TestRun (22.00s)
PASS
Too many goroutines running after all test(s).
1 instances of:
net.runtime_pollWait(...)
	/usr/local/go/src/runtime/netpoll.go:160 +0x60
net.(*pollDesc).Wait(...)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(...)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).accept(...)
	/usr/local/go/src/net/fd_unix.go:426 +0x27c
net.(*TCPListener).AcceptTCP(...)
	/usr/local/go/src/net/tcpsock_posix.go:254 +0x4d
net/http.tcpKeepAliveListener.Accept(...)
	/usr/local/go/src/net/http/server.go:2425 +0x41
net/http.(*Server).Serve(...)
	/usr/local/go/src/net/http/server.go:2117 +0x129
net/http.(*Server).ListenAndServe(...)
	/usr/local/go/src/net/http/server.go:2098 +0x136
net/http.ListenAndServe(...)
	/usr/local/go/src/net/http/server.go:2195 +0x98
github.com/coreos/etcd/tools/functional-tester/etcd-tester.Run.func1()
	/home/gyuho/go/src/github.com/coreos/etcd/tools/functional-tester/etcd-tester/main.go:57 +0x3c
created by github.com/coreos/etcd/tools/functional-tester/etcd-tester.Run
	/home/gyuho/go/src/github.com/coreos/etcd/tools/functional-tester/etcd-tester/main.go:57 +0x7f0
1 instances of:
net.runtime_pollWait(...)
	/usr/local/go/src/runtime/netpoll.go:160 +0x60
net.(*pollDesc).Wait(...)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(...)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(...)
	/usr/local/go/src/net/fd_unix.go:250 +0x23a
net.(*conn).Read(...)
	/usr/local/go/src/net/net.go:172 +0xe4
bufio.(*Reader).fill(...)
	/usr/local/go/src/bufio/bufio.go:97 +0x1e9
bufio.(*Reader).Read(...)
	/usr/local/go/src/bufio/bufio.go:207 +0x260
io.ReadAtLeast(...)
	/usr/local/go/src/io/io.go:297 +0xe6
io.ReadFull(...)
	/usr/local/go/src/io/io.go:315 +0x62
encoding/gob.decodeUintReader(...)
	/usr/local/go/src/encoding/gob/decode.go:119 +0x92
encoding/gob.(*Decoder).recvMessage(...)
	/usr/local/go/src/encoding/gob/decoder.go:76 +0x5e
encoding/gob.(*Decoder).decodeTypeSequence(...)
	/usr/local/go/src/encoding/gob/decoder.go:140 +0x47
encoding/gob.(*Decoder).DecodeValue(...)
	/usr/local/go/src/encoding/gob/decoder.go:208 +0x15d
encoding/gob.(*Decoder).Decode(...)
	/usr/local/go/src/encoding/gob/decoder.go:185 +0x289
net/rpc.(*gobClientCodec).ReadResponseHeader(...)
	/usr/local/go/src/net/rpc/client.go:223 +0x51
net/rpc.(*Client).input(...)
	/usr/local/go/src/net/rpc/client.go:109 +0xbf
created by net/rpc.NewClientWithCodec
	/usr/local/go/src/net/rpc/client.go:201 +0xd7
18 instances of:
time.Sleep(...)
	/usr/local/go/src/runtime/time.go:59 +0xf9
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc.(*Conn).resetTransport(...)
	/home/gyuho/go/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/clientconn.go:462 +0x8e4
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc.(*Conn).reconnect(...)
	/home/gyuho/go/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/clientconn.go:497 +0xc4
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc.(*Conn).transportMonitor(...)
	/home/gyuho/go/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/clientconn.go:522 +0x16b
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc.NewConn.func1(...)
	/home/gyuho/go/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/clientconn.go:322 +0x1b5
created by github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc.NewConn
	/home/gyuho/go/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/clientconn.go:323 +0x506
exit status 1
FAIL	github.com/coreos/etcd/tools/functional-tester/etcd-tester	22.013s

```

File: tools/functional-tester/etcd-tester/stresser.go [hunk 1]
```diff
@@ -103,11 +115,11 @@ func (s *stresser) Stress() error {
 
 func (s *stresser) Cancel() {
 	s.mu.Lock()
-	cancel, conn := s.cancel, s.conn
+	cancel, conn, wg := s.cancel, s.conn, s.wg
 	s.mu.Unlock()
 	cancel()
-	// TODO: wait for all routines to exit by adding a waitGroup before calling conn.Close()
 	conn.Close()
+	wg.Wait()
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/stresser.go]: can you follow the todo? close the conn after wg.Wait? then we probably do not have to check a lot of connection errros.

File: tools/functional-tester/etcd-tester/tester.go [hunk 2]
```diff
@@ -215,6 +215,7 @@ func (c *cluster) getRevision() (map[string]int64, error) {
 		if err != nil {
 			return nil, err
 		}
+		defer conn.Close()
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/tester.go]: do not use defer in a for loop. let's close right after we finish using it.

File: tools/functional-tester/etcd-tester/tester.go [hunk 3]
```diff
@@ -234,6 +235,7 @@ func (c *cluster) getKVHash() (map[string]int64, error) {
...
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/tester.go]: same here.

File: tools/functional-tester/etcd-tester/tester.go [hunk 4]
```diff
@@ -256,6 +258,7 @@ func (c *cluster) compactKV(rev int64) error {
 		if err != nil {
 			continue
 		}
+		defer conn.Close()
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/tester.go]: same here.

File: tools/functional-tester/etcd-tester/tester.go (hunk 2 above)
Comment (gyuho) [inline tools/functional-tester/etcd-tester/tester.go]: Ok

File: tools/functional-tester/etcd-tester/stresser.go [hunk 5]
```diff
@@ -83,7 +92,10 @@ func (s *stresser) Stress() error {
 					Value: []byte(randStr(s.KeySize)),
 				})
 				putcancel()
-				if grpc.ErrorDesc(err) == context.Canceled.Error() {
+				if grpc.ErrorDesc(err) == context.Canceled.Error() ||
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/stresser.go]: if we really need so many tests, we should make this a function.

File: tools/functional-tester/etcd-tester/stresser.go [hunk 6]
```diff
@@ -65,17 +66,25 @@ func (s *stresser) Stress() error {
...
 	s.conn = conn
 	s.cancel = cancel
+	s.wg = wg
 	s.mu.Unlock()
 
 	kvc := pb.NewKVClient(conn)
 
 	for i := 0; i < s.N; i++ {
 		go func(i int) {
+			defer func() {
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/stresser.go]: defer wg.Done()

File: tools/functional-tester/etcd-tester/stresser.go (hunk 5 above)
Comment (gyuho) [inline tools/functional-tester/etcd-tester/stresser.go]: Ok will make it into a function. (just tested and checked that we need to check all those)

File: tools/functional-tester/etcd-tester/stresser.go [hunk 7]
```diff
@@ -83,7 +92,10 @@ func (s *stresser) Stress() error {
 					Value: []byte(randStr(s.KeySize)),
 				})
 				putcancel()
-				if grpc.ErrorDesc(err) == context.Canceled.Error() {
+				if grpc.ErrorDesc(err) == context.Canceled.Error() ||
+					grpc.ErrorDesc(err) == context.DeadlineExceeded.Error() ||
+					grpc.ErrorDesc(err) == "transport is closing" ||
+					grpc.ErrorDesc(err) == "grpc: the client connection is closing" {
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/stresser.go]: use grpc.ErrConnClosing.

File: tools/functional-tester/etcd-tester/stresser.go (hunk 5 above)
Comment (xiang90) [inline tools/functional-tester/etcd-tester/stresser.go]: probably we should just return if there is an error...

**Comment by gyuho** [comment]
@xiang90 PTAL. Thanks! Will investigate grpc leaky goroutines separately.

File: tools/functional-tester/etcd-tester/stresser.go [hunk 8]
```diff
@@ -83,7 +90,7 @@ func (s *stresser) Stress() error {
 					Value: []byte(randStr(s.KeySize)),
 				})
 				putcancel()
-				if grpc.ErrorDesc(err) == context.Canceled.Error() {
+				if checkCancelErr(err) {
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/stresser.go]: how about just removing this func and return on error? also we should get rid of the fail count at line 98.

Comment (gyuho) [inline tools/functional-tester/etcd-tester/stresser.go]: Sure. Will return when err and increase the failure count before return.

Comment (xiang90) [inline tools/functional-tester/etcd-tester/stresser.go]: we probably do not want to count failure anymore. previously we only have one long running stresser, so we want to count the overall failure/success. now, when there is a failure we will always return. so it is pointless to do the accounting anymore.

File: tools/functional-tester/etcd-tester/stresser.go [hunk 9]
```diff
@@ -83,15 +90,14 @@ func (s *stresser) Stress() error {
...
-					return
-				}
-				s.mu.Lock()
 				if err != nil {
+					s.mu.Lock()
 					s.failure++
-				} else {
-					s.success++
+					s.mu.Unlock()
+					return
```
Comment (gyuho) [inline tools/functional-tester/etcd-tester/stresser.go]: @xiang90 Just added these lines.

File: tools/functional-tester/etcd-tester/stresser.go [hunk 10]
```diff
@@ -103,17 +100,18 @@ func (s *stresser) Stress() error {
...
 	conn.Close()
 }
 
-func (s *stresser) Report() (success int, failure int) {
+func (s *stresser) Report() (int, int) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
-	return s.success, s.failure
+	// TODO: find a better way to report v3 tests
+	return -1, -1
```
Comment (gyuho) [inline tools/functional-tester/etcd-tester/stresser.go]: Just added todo here.

**Comment by xiang90** [comment]
lgtm

**Comment by gyuho** [comment]
thanks. Will deploy with the new tester.

**Pull request merged**